core/src/com/unciv/ui/overviewscreen/EspionageOverviewScreen.kt

```diff
@@ -0,0 +1,166 @@
+import com.unciv.ui.utils.extensions.addSeparator
+import com.unciv.ui.utils.extensions.addSeparatorVertical
+import com.unciv.ui.utils.extensions.disable
+import com.unciv.ui.utils.extensions.enable
+import com.unciv.ui.utils.extensions.onClick
+import com.unciv.ui.utils.extensions.setSize
+import com.unciv.ui.utils.extensions.toLabel
+import com.unciv.ui.utils.extensions.toTextButton
+
+/** Screen used for moving spies between cities */
+class EspionageOverviewScreen(val civInfo: CivilizationInfo) : PickerScreen(true) {
+    private val collator = UncivGame.Current.settings.getCollatorFromLocale()
+
+    private val spySelectionTable = Table(skin)
+    private val spyScrollPane = AutoScrollPane(spySelectionTable)
+    private val citySelectionTable = Table(skin)
+    private val cityScrollPane = AutoScrollPane(citySelectionTable)
+    private val headerTable = Table(skin)
+    private val middlePanes = Table(skin)
+
+    private var selectedSpyButton: Button? = null
+    private var selectedSpy: Spy? = null
+    // if the value == null, this means the Spy Hideout.
+    private var moveSpyHereButtons = hashMapOf<Button,CityInfo?>()
+
+    init {
+        topTable.add(headerTable)
+        topTable.addSeparator()
+
+        middlePanes.add(spyScrollPane)
+        middlePanes.addSeparatorVertical()
+        middlePanes.add(cityScrollPane)
+        topTable.add(middlePanes)
+
+        update()
+
+        closeButton.isVisible = true
+        setDefaultCloseAction()
+    }
+
+    private fun update() {
+        updateSpyList()
+        updateCityList()
+    }
+
+    private fun updateSpyList() {
+        spySelectionTable.clear()
+        spySelectionTable.add("Spy".toLabel()).pad(5f)
+        spySelectionTable.add("Location".toLabel()).pad(5f).row()
+        for (spy in civInfo.espionageManager.spyList) {
+            spySelectionTable.add(spy.name.toLabel()).pad(5f)
+            spySelectionTable.add(spy.getLocationName(civInfo.gameInfo)).pad(5f)
+
+            val moveSpyButton = "Move".toTextButton()
+            moveSpyButton.onClick {
+                resetSelection()
+                selectedSpyButton = moveSpyButton
+                selectedSpy = spy
+                moveSpyButton.disable()
+                for ((button, city) in moveSpyHereButtons)
+                    // For now, only allow spies to be send to cities of other major civs
+                    // Not own cities as counterintelligence isn't implemented
+                    // Not city-state civs as rigging elections isn't implemented
+                    // Technically, stealing techs from other civs also isn't implemented, but its the first thing I'll add so this makes the most sense to allow.
+                    if (city == null || ( // hideout
+                        city.civInfo.isMajorCiv()
+                        && city.civInfo != civInfo
+                        && (!city.espionage.hasSpyOf(civInfo) || selectedSpy!!.location == city.id)
```

Comment:
`(!city.espionage.hasSpyOf(civInfo) || selectedSpy!!.location == city.id)` - it is always true, because `city.espionage.hasSpyOf(civInfo)` does the same check `selectedSpy!!.location == city.id`.
I am not sure `||` is placed correctly here.
Also, `city == null` allows to move a spy to the hideout even if it is already there. Is it expected behaviour?

Reply:
Not exactly, `!city.espionage.hasSpyOf(civInfo)` means the spy can move to any city not containing a spy, or `selectedSpy!!.location == city.id` to the location the spy is currently stationed (if you don't want to move it even though you selected it)

Spies should always be able to move to the hideout, moving it while its already there effectively cancels the movement

Reply:
Ok, then I misunderstand the UI.
I would suggest cancelling the movement by pressing the same button "Move" which is highlighted (or renamed to "Cancel") until the "<=" button is pressed.
The possibility to move the spy to the location where it is already present seems to me like a nonsense.